Transfer: Transactions Review

Soaryn opened the pull request:
The transaction slice of PR #1115

Soaryn committed:
Transfer: Transactions

Soaryn committed:
Adds missing deprecation handling file from slice

covers1624 commented on src/main/java/net/neoforged/neoforge/transfer/transaction/UnsafeTransactionManager.java:
```diff
@@ -0,0 +1,51 @@
+ * In short, <strong>use this as a last resort</strong>.
+ */
+@ApiStatus.Internal
+public final class UnsafeTransactionManager {
```
I'm not happy with the entire concept of these 'Unsafe' utilities.

Are there actual use cases for these in further slices? Can you link them?

Soaryn replied:
Regardless of slices, there are scenarios where we are unable to pass in the current transaction to a method, (typically because the method is not in the modder's control so they can't change the signature) say like `getCapacity` or `getAmount`. Let's take a pipe example that would want to inquire on some `insert` method of a connected handler group to report that value in `getCapacity`. Being able to open a transaction from the current level allows them to safely "simulate" that without breaking the transaction system.

As for the slices: EnergyHandlerUtil and ResourceHandlerUtil only really uses them for inquiry based returns which will never commit. The migration wrapped resource handlers use them (however, I don't really think we should keep the old handlers functional).

covers1624 replied:
Then we should push the system towards not requiring the parent transaction context _at all_ when a new transaction is constructed. It should just be picked up automatically when you do `try (var transaction = Transaction.open())`.

Technici4n replied:
I think we should always pass the transaction explicitly. `getCurrentOpenedTransaction` is fine (but there is no need to move it to a separate class), but `openUnsafe` doesn't make sense to me.

Opening a transactions from `getCapacity` does not make sense to me, and will probably break the transaction system.

covers1624 replied:
I'll trust @Technici4n's judgment here, as he has more experience in this field than me.

We should definitely just provide `Transaction.getCurrentTransaction()` or similar.

Soaryn replied:
I've moved the `getCurrentOpenedTransaction()` to `Transaction`, but the class is left in as a temporary measure during the other slices (at least until the wrappers as noted in the file) as it will make in lining much simpler to do throughout the other slices when used. It will be removed in the end, but I'd at least like to keep it around to remove the unnecessary work burden introduced by the slicing.

> getCapacity does not make sense to me

In regards to the example of opening a transaction in getCapacity: My pipes' capacities are based on the current "demand" of the connected system. They themselves don't have an innate capacity, but rather a dynamic value that is allowed to be accepted. This approach works rather well, but when inquired "how much can you store?" (`getCapacity`) I need to do a lookup once in that tick to determine an estimation using `insert`. Since it is in `getCapacity`, I can't just open a root transaction as there is no guarantee it isn't already in one, hence the `openUnsafe` (or now `open(getCurrentOpenedTransaction())`). I have already tested it as that is what lead to it being readded. It would only really break if it couldn't properly roll back, which is a separate issue which would happen anywhere else.

covers1624 replied:
> but the class is left in as a temporary measure during the other slices

The whole idea behind the slices is to get each slice reviewed, and finalized. No files included in a slice should be modified by a later slice.

Soaryn replied:
That wont happen anyway. We already have some methods asking to be "delayed" into a different slice, which would force classes to be written. ResourceHandlerUtil was fractured to handle the Resource slice

Slicing effectively added a lot of refactoring burdens that would never be present otherwise (which is what I want to try to avoid as I can just inline with IDE when those uses are present) this class would only change one more tike later when it is removed. What you are wanting is far too idealistic. It'd be different if this was brand new code we were building off of, but this is a compartamentalization of what is already there. 

To be clear: the removal of this particular file would be in its own slice not in the slice it was used to inline the methods. There will be a slice dedicated to cleaning up in lined files. Probably after the "misc" slice.

covers1624 replied:
This is not how reviews work.

We are requesting that you make these changes, the refactoring burden is on you, yes. I have even suggested tools which would reduce this burden (GitButler), you could have all the slices checked out at the same time locally, reducing this burden to effectively zero.

Regardless, if we are going to keep having to fight every step of the way for changes, I personally don't see the point in continuing with the transfer api at all.

Soaryn replied:
What I am saying is, I have been asked to do this for the Resources slice, and you are telling me not to do this for this slice? It seems like I am getting conflicting info as a problem.

Technici4n replied:
The point is that each slice should stand on its own. So please remove all unused/legacy stuff. Future slices have not been properly reviewed yet and are not guaranteed to use any of it.

covers1624 commented on src/main/java/net/neoforged/neoforge/transfer/transaction/TransactionManager.java:
```diff
@@ -0,0 +1,114 @@
+     * @throws IllegalStateException If a parent is passed, but it's not the current transaction.
+     * @throws IllegalStateException If a parent is passed, but it was already closed.
+     */
+    public static Transaction open(@Nullable TransactionContext parent) {
```
I would personally prefer if this was on `Transaction` directly (even if it calls this method). It meshes better with other similar functions `Optional.of`, `List.of`, `ItemResource.of`, etc.

covers1624 commented on src/main/java/net/neoforged/neoforge/transfer/transaction/TransactionManager.java:
```diff
@@ -0,0 +1,114 @@
+    /**
+     * @return The current lifecycle of the transaction stack on this thread.
+     */
+    public static Transaction.Lifecycle getLifecycle() {
```
Much the same as my comment regarding the Unsafe utilities.

What uses cases do we have for this in later slices?

Soaryn replied:
This can be used to allow someone to respond differently to changes based on if they are in a transaction or not. I use it in XyCraft. Being able to know if a transaction is live seems like perfectly valid date to have.

as for uses in later slices, we don't have any, but this is a limited window scope of getting the current thread's action cycle. No reason to not expose it.

covers1624 replied:
What do you use it for specifically, can you show some code? I don't see any reason why something would need to know it's inside a transaction, and if they do, the transaction should be provided.

Soaryn replied:
There are a number of ways to solve this. This way is viable:
![image](https://github.com/user-attachments/assets/5324643c-98eb-4551-b5fb-f0e15e3fb6df)

covers1624 replied:
Fair enough, I guess some internals of containers, etc, may require knowledge about the current transaction for whatever record keeping/`setChanged` they may have.

I would suggest that these functions be moved to `Transaction`. Perhaps they should be instance methods on transaction, `Transaction.getLifecycle()` and `Transaction.isActive.`, Your call would then change to `!Transaction.getCurrentTransaction().isActive()`, Although nullable makes that annoying. Happy for the method to have a name of `Transaction.isCurrentTransactionActive()` which does the appropriate guards.

Technici4n replied:
This was originally added in https://github.com/FabricMC/fabric/pull/2086 because on Fabric `Transaction.getCurrentUnsafe()` would throw IllegalStateException (ISE) if invoked from a close or outer close callback, and having an enum that clearly represents the current lifecycle was better than catching the ISE. I think we should restore the ISE throw in `getCurrentTransaction`, as the transaction should not be accessible at all during the callbacks, even from the "unsafe" method. This also makes the method not-so-unsafe in the end. 😉

covers1624 commented on src/main/java/net/neoforged/neoforge/transfer/transaction/TransactionManager.java:
```diff
@@ -0,0 +1,114 @@
+    final List<TransactionContext.RootCloseCallback> rootCloseCallbacks = new ArrayList<>();
+    @ApiStatus.Internal
+    int currentDepth = -1;
+    @ApiStatus.Internal
```
All these fields are already package private. No one can package squat them to access them (modules). They don't need to be marked as internal.

Same with the functions below this comment too.

covers1624 commented on src/main/java/net/neoforged/neoforge/transfer/transaction/Transaction.java:
```diff
@@ -0,0 +1,245 @@
+        String self = TransactionManager.debugNameFrom(manager.debugMap.get(nestingDepth));
+        String actual = TransactionManager.debugNameFrom(manager.debugMap.get(manager.currentDepth));
+
+        CrashReport report = CrashReport.forThrowable(new IllegalStateException("Transaction function was called on a transaction with depth %d, but the current transaction has depth %d."
```
Whilst this reported exception is neat. I don't really think it's necessary, a simpler exception stating that child transactions are still open should be sufficient.

We highly promote the use of `AutoCloseable` with Transactions, so this should not really be an issue.

We could perhaps have some sysprop or log filter for devs to enable, which stack prints when transactions are created, for debugging this stuff.

Technici4n replied:
In general, using a reported exception probably only makes sense when catching an existing exception.

covers1624 commented on src/main/java/net/neoforged/neoforge/transfer/transaction/TransactionContext.java:
```diff
@@ -0,0 +1,122 @@
+public interface TransactionContext {
+    /**
+     * @return The nesting depth of this transaction: 0 if it has no parent,
+     *         1 if its parent has no parent, and so on...
```
`1 if its parent has no parent, and so on...` is not really descriptive enough.

Clearer docs would be that this method returns the distance away from the root node.

Shadows-of-Fire commented on src/main/java/net/neoforged/neoforge/transfer/transaction/TransactionManager.java:
```diff
@@ -0,0 +1,114 @@
+ *
+ * @see Transaction
+ */
+public final class TransactionManager {
```
Based on intention, this entire class should be `@ApiStatus.Internal`. Any according APIs should be done through the `Transaction` class. See comments from @covers1624 regarding the request for `Transaction.open()` and `Transaction.getCurrentTransaction()`

Shadows-of-Fire commented on src/main/java/net/neoforged/neoforge/transfer/transaction/Transaction.java:
```diff
@@ -0,0 +1,245 @@
+        if (manager.currentDepth != -1 && manager.stack.get(manager.currentDepth) == this)
+            return;
+
+        //TODO validate this is handling the use case of showing a dev which transactions are being opened / closed (on top of the stacktrace)
```
This TODO should be resolved as part of this PR. Presumably via test code.

Shadows-of-Fire commented on src/main/java/net/neoforged/neoforge/transfer/transaction/Transaction.java:
```diff
@@ -0,0 +1,245 @@
+        // That is why every callback has to run inside their own try-with-resources block.
+        RuntimeException closeException = null;
+
+        // Invoke callbacks in reverse order
```
This comment should state why we're invoking them in reverse order.

Soaryn replied:
@Technici4n, can you remind me why we do so in reverse? Or perhaps have a better comment we could put here?

Shadows-of-Fire commented on src/main/java/net/neoforged/neoforge/transfer/transaction/Transaction.java:
```diff
@@ -0,0 +1,245 @@
+        if (manager.currentDepth == 0) {
+            lifecycle = Lifecycle.ROOT_CLOSING;
+
+            // Invoke outer close callbacks in reverse order
```
This comment should state why we're invoking them in reverse order.

Soaryn replied:
@Technici4n, can you remind me why we do so in reverse? Or perhaps have a better comment we could put here?

Technici4n replied:
It's just how it was always done, but I'm not sure it makes more sense than doing it in the normal order, especially now that callbacks cannot be registered directly anymore.

Soaryn replied:
From what I can tell the order is agnostic to the actions really. So unless a reason is found, we'll just iterate forwards with an enhanced loop. Seems to at least validate the old tests.

Shadows-of-Fire commented on src/main/java/net/neoforged/neoforge/transfer/transaction/Transaction.java:
```diff
@@ -0,0 +1,245 @@
+ * <li>Calling {@link TransactionManager#open} with a non-{@code null} parent creates a new nested transaction, i.e. a new checkpoint with the current state.
+ * Committing a nested transaction will validate the changes that happened, but they may
+ * still be cancelled later if a parent transaction is cancelled.
+ * Aborting a nested transaction immediately reverts the changes - cancelling any modification made after the call
+ * to {@link TransactionManager#open}.</li>
```
```suggestion
 * Aborting a nested transaction immediately reverts the changes within that nested transaction - cancelling any modification made after the call
 * to {@link TransactionManager#open}.</li>
```

Shadows-of-Fire commented:
The docs in some classes are inconsistent with the naming conventions being used for [inner/outer, root/nested, parent/child, etc) transactions. While it isn't totally relevant which variant pair we select, we need to consistently select one.

It seems like, in general, we would prefer to use `root` for the base layer (i.e. `open(null)` and anything bound to that). Given that, I suggest we not use `nested` for _anything_, and use `inner` / `outer` for referring to generic discussions where multiple transactions are at play.

Soaryn committed:
Update src/main/java/net/neoforged/neoforge/transfer/transaction/Transaction.java

Changes documentation to be more clear in intent of transaction aborting consequences.

Co-authored-by: Brennan Ravenscroft <Bward7864@gmail.com>

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/transaction/SnapshotJournal.java:
```diff
@@ -0,0 +1,171 @@
+ * is retained. {@link #onCommit} will be called after the transaction is closed
+ * and then {@link #releaseSnapshot} will be called because the snapshot is not necessary anymore.
+ *
+ * @param <T> The objects that this participant uses to save its state snapshots.
```
participant is not used anymore, the wording should be updated

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/transaction/SnapshotJournal.java:
```diff
@@ -0,0 +1,171 @@
+ * to signal that the current state should revert to that of the snapshot.
+ * The snapshot object is then {@linkplain #releaseSnapshot released}, and can be cached for subsequent use, or discarded.
+ *
+ * <p>When an outer transaction is committed, {@link #revertToSnapshot} will not be called so that the current state of this participant
```
we don't use "outer" anymore so the wording should be updated

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/transaction/SnapshotJournal.java:
```diff
@@ -0,0 +1,171 @@
+    @Nullable
+    @Deprecated(since = ResourceHandlerDeprecationHandling.MC_1_21_6)
+    private static SnapshotJournal<?> DEEPEST_SNAPSHOT = null;
+    private final ArrayList<T> snapshots = new ArrayList<>();
```
```suggestion
    private final List<T> snapshots = new ArrayList<>();
```

Soaryn replied:
We do an ensureCapacity, so not able to do that per se. Nor really is there a reason to.

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/transaction/SnapshotJournal.java:
```diff
@@ -0,0 +1,171 @@
+     * @param originalState state of this participant before the transactional operation.
+     *                      This corresponds to the first {@link #createSnapshot() snapshot} that was created in the transactional operation.
+     */
+    protected void onCommit(T originalState) {}
```
This should be called `onFinalCommit`/`onRootCommit` to make it clear that it only applies to the root transaction being committed.

Soaryn replied:
I'll put `onRootCommit(T)` for now, but when looking over it more recently, I was actually considering doing `revertTo(T)` and `commitFrom(T)` as possible options given the documentation even states when these would occur.

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/transaction/SnapshotJournal.java:
```diff
@@ -0,0 +1,171 @@
+     * Called after an outer transaction succeeded,
+     * to perform irreversible actions such as {@code setChanged()} or neighbor updates.
+     * <p>
+     * Assume that without any implementation, data on something like BlockEntities that expect to be told when its data is written to, could be potentially lost on world save.
```
Kinda weird comment, I would just remove it as the wording is not great.

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/transaction/SnapshotJournal.java:
```diff
@@ -0,0 +1,171 @@
+     * to perform irreversible actions such as {@code setChanged()} or neighbor updates.
+     * <p>
+     * Assume that without any implementation, data on something like BlockEntities that expect to be told when its data is written to, could be potentially lost on world save.
+     *
```
Add a note saying that opening a transaction from this function will typically throw as the current one is not fully closed yet.

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/transaction/SnapshotJournal.java:
```diff
@@ -0,0 +1,171 @@
+    /**
+     * Update the stored snapshots so that the changes happening as part of the passed transaction can be correctly
+     * committed or rolled back.
+     * This function should be called every time the participant is about to change its internal state as part of a transaction.
```
We don't use participant anymore

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/transaction/SnapshotJournal.java:
```diff
@@ -0,0 +1,171 @@
+ *
+ * @param <T> The objects that this participant uses to save its state snapshots.
+ */
+public abstract class SnapshotJournal<T> implements Transaction.CloseCallback, TransactionContext.RootCloseCallback {
```
If we want to we can make this support `null` snapshots by masking nulls. This would also simplify the cases where we don't actually need the snapshot, and `Void` would be the most appropriate type. I can send a PR to this PR.

Soaryn replied:
Looking into a bit, making it have null support causes some of the internals to go a little out of sync for not much gain. We can just make a custom object type like in `StatelessSnapshotJournal` as that is precisely what is solved with it without reworking the system.

While `SnapshotJournal<Void>` looks cool, it does cause some headaches elsewhere since it now needs to support nullability. Not sure it is worth the one off instance to make one type be nullable where all others are expected to be non-null.

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/transaction/SnapshotJournal.java:
```diff
@@ -0,0 +1,171 @@
+    @Override
+    public void onClose(TransactionContext transaction, Transaction.Result result) {
+        //region Migration phase removal
+        //Neo: for testing and will be removed after deprecation period is over for handler reworks.
```
Don't use the `Neo:` comment prefix in NeoForge classes - it's exclusively for use in patches.

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/transaction/SnapshotJournal.java:
```diff
@@ -0,0 +1,171 @@
+    public void onClose(TransactionContext transaction, Transaction.Result result) {
+        //region Migration phase removal
+        //Neo: for testing and will be removed after deprecation period is over for handler reworks.
+        // This is to provide a quick way to give some metrics during the migration phase
+        int currentDepth = transaction.nestingDepth();
+        int max = Math.max(DEEPEST_LAYER, currentDepth);
+        if (max != DEEPEST_LAYER) {
+            DEEPEST_LAYER = max;
+            DEEPEST_SNAPSHOT = this;
+        }
+        //endregion
```
Honestly I don't see what this information will tell us, so I would lean towards removal. Saving the deepest snapshot might also be problematic wrt. memory leaks.

Soaryn replied:
Really all we need from this is the class name (package included) when the max changes, so the latter part is solvable by just keeping that instead of the object. However, for what the information can tell us, unless you'd like to make a full blown diagnostic tool for this, this would at least help isolate when something is rather large in the initial migrations to help confirm assumptions that we have made on how deep a layer would get. You assume fairly low, but I am fairly confident the layer depth will actually get quite large. This was a way to confirm it a bit without causing any unnecessary grief. 

If you know of a better way I am all ears. This isn't intended to be persistent past a full version, just the initial migrations.

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/transaction/SnapshotJournal.java:
```diff
@@ -0,0 +1,171 @@
+        T snapshot = snapshots.remove(currentDepth);
+
+        //If the transaction was aborted, revert to snapshot
+        if (result.wasAborted()) {
```
Still not as clear as my original, which you should just copy-paste without making further modifications: https://github.com/neoforged/NeoForge/blob/fd9650c9dcd30b35725f5ca9aa1cf9e8f84cea0c/src/main/java/net/neoforged/neoforge/transfer/transaction/SnapshotParticipant.java#L99

FiniteReality replied:
I disagree; personally the code currently here is clearer.

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/transaction/SnapshotJournal.java:
```diff
@@ -0,0 +1,171 @@
+        }
+    }
+
+    /**
+     * @return The deepest nested layer from any transaction over the lifetime of the runtime. This is intended to identify some possible changes needed after migration. Not used outside Neo.
+     */
+    @Deprecated(since = ResourceHandlerDeprecationHandling.MC_1_21_6)
+    @ApiStatus.Internal
+    public static int getDeepestLayer() {
+        return DEEPEST_LAYER;
+    }
+
+    /**
+     * @return The deepest nested layer from any transaction over the lifetime of the runtime. This is intended to identify some possible changes needed after migration. Not used outside Neo.
+     */
+    @Deprecated(since = ResourceHandlerDeprecationHandling.MC_1_21_6)
+    @ApiStatus.Internal
+    public static String getDeepestSnapshot() {
+        if (DEEPEST_SNAPSHOT == null) return "Nothing";
+        return DEEPEST_SNAPSHOT.getClass().toString();
+    }
```
Just very weird overall. I'm not sure what information we expect to gain from this.

Soaryn replied:
Well the original reason: Your assumption is that the depth won't get very high, "maybe 2-3", but when you said that, we had a test that was already achieving 5 in depth after removing a couple unnecessary layers. So in part, this was to confirm during initial migrations just large the depth will get. As the snapshot array list's memory cost is based on capacity and object size, not list size. It costs nothing for us to really have (the deepest snapshot we can just store the string instead of resolving at the end) for a couple iterations for validations.

If you know of a better way to confirm the use then I am all ears.

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/transaction/Transaction.java:
```diff
@@ -0,0 +1,245 @@
+import net.minecraft.ReportedException;
+
+/**
+ * A global operation where participants guarantee atomicity: either the whole operation succeeds,
```
```suggestion
 * A global operation that guarantees atomicity: either the whole operation succeeds,
```

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/transaction/Transaction.java:
```diff
@@ -0,0 +1,245 @@
+ * to give every participant a chance to react to transaction close events.
+ *
+ * <p>This is very low-level for most applications, and most participants should subclass {@link SnapshotJournal}
+ * that will take care of properly maintaining their state.
```
We should reword this to only link to SnapshotJournal, and probably make the callbacks internal details. See below.

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/transaction/Transaction.java:
```diff
@@ -0,0 +1,245 @@
+ * that will take care of properly maintaining their state.
+ *
+ * <p>Journals should generally be passed a {@link TransactionContext} parameter instead of the full {@code Transaction},
+ * to make sure they don't accidentally call {@link #commit} or {@link #close}.
```
This is not exclusive to journals. In general there is no reason to pass a `Transaction` around.

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/transaction/TransactionContext.java:
```diff
@@ -0,0 +1,122 @@
+@ApiStatus.NonExtendable
+public interface TransactionContext {
+    /**
+     * @return The nesting depth of this transaction: 0 if it has no parent,
```
We should avoid having a `@return` tag as the only javadoc because it means that no summary will be rendered in the javadoc html view.

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/transaction/TransactionContext.java:
```diff
@@ -0,0 +1,122 @@
+     *
+     * @throws IllegalStateException If this function is not called on the thread this transaction was opened in.
+     */
+    void addCloseCallback(CloseCallback closeCallback);
```
I think we should remove the callbacks from the API and instead make `SnapshotJournal` the only exposed way of interacting with transactions. The original intention was that there would be many kinds of participants and the snapshot journal (originally called `SnapshotParticipant`) would just be one of them. In practice however it turns out that the `SnapshotParticipant` is all you need, and other uses of callbacks are typically mistakes by mod authors not familiar with the transaction system.

Soaryn replied:
Just to be clear you want the close callback methods and the functional interfaces removed correct?

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/transaction/TransactionContext.java:
```diff
@@ -0,0 +1,122 @@
+    /**
+     * The result of a transaction operation.
+     */
+    enum Result {
```
If we remove the callbacks this also becomes internal.

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/transaction/Transaction.java:
```diff
@@ -0,0 +1,245 @@
+        if (manager.currentDepth == 0) {
+            lifecycle = Lifecycle.ROOT_CLOSING;
+
+            // Invoke outer close callbacks in reverse order
```
```suggestion
            // Invoke root close callbacks in reverse order
```

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/transaction/Transaction.java:
```diff
@@ -0,0 +1,245 @@
+        lifecycle = Lifecycle.CLOSING;
+
+        // Note: it is important that we don't let exceptions corrupt the global state of the transaction manager.
+        // That is why every callback has to run inside their own try-with-resources block.
```
```suggestion
        // That is why every callback has to run inside its own try block.
```

Soaryn replied:
We likely want to keep the "with resources" part just to help educate or provide what to search for when they go look this up should they need to.

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/transaction/snapshots/IndexedIntSnapshot.java:
```diff
@@ -0,0 +1,101 @@
+ * that can be updated when an index gets written to. An example is using a {@link NotifyingSnapshotJournal} to handle
+ * the scenario of doing a single commit for something like {@link BlockEntity#setChanged()}.
+ */
+public class IndexedIntSnapshot extends SnapshotJournal<Integer> {
```
This I would remove, it is a lot of indirection just to store an int index. Better have a dedicated subclass where it matters.

Technici4n replied:
Remember to remove this.

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/transaction/snapshots/NotifyingSnapshotJournal.java:
```diff
@@ -0,0 +1,92 @@
+ * <p>
+ * When the transaction is committed, {@link #onCommit} will run; and when the transaction is reverted {@link #onRevert} will instead. Only one per transaction chain will be called and only once.
+ */
+public final class NotifyingSnapshotJournal extends SnapshotJournal<NotifyingSnapshotJournal.IgnoredValue> {
```
This should be removed as well in favor of just overriding `SnapshotJournal` where it's needed, probably with a `Void` snapshot type if we do the null masking. @pupnewfster and I already asked to remove base "snapshots" from the previous PR, which you did. The intention was of course NOT to add back even more complex versions of them later.

Soaryn replied:
The indexed one we **can** remove, though it doesn't serve much point to fully given it would be a simple to reuse journal for anyone else's energy handler journal. 

The notifying one is something I do recommend we keep as that is what you want from the `Void` variant, without having to restructure things to support an edge-cased `null`. It really isn't complex. They serve a purpose, and while we could have everyone make a dedicated subclass, there isn't a lot of reason to when they are going to be virtually the same in many scenarios such as the intended ones these are solving for.

Arguably the indexable int one was designed around indexable energy which was removed, but the backing structure being able to expand back into that was relatively weightless.

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/transaction/Transaction.java:
```diff
@@ -0,0 +1,245 @@
+ * <p>This is illustrated in the following example.
+ *
+ * <pre>{@code
+ * try (Transaction outerTransaction = TransactionManager.open(null)) {
```
outer should be replaced by root

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/transaction/TransactionContext.java:
```diff
@@ -0,0 +1,122 @@
+    void addRootCloseCallback(RootCloseCallback rootCloseCallback);
+
+    /**
+     * A callback that is invoked after the outer transaction is closed.
```
root

Soaryn committed:
Changes migration debugging info into SnapshotJournalDebugInfo to present a little more information, as well as make it cleaner/simpler to remove post migration.

Removes callback methods from TransactionContext. This all should now be handled by SnapshotJournals on the consumer side should they need more handling. Still may be things we can clean up about this.

Moved `Result` to Transaction and made it private/

Moves static methods from TransactionManager to Transaction to match style of things like `List.of`

Removes unnecessary Internal annotations on package private members

Adds a StatelessSnapshot base that is equivalent in use to `SnapshotJournal<Void>` while keeping the infrastructure the same. An attempt at making SnapshotJournal<Void> was made.

Changes name of `nestingDepth` to just `depth` as it was redundant. In this case depth is always nested in the context of a transaction.

Handles wording of documentation to try to keep it consistent. There are still places to improve intent.

Adds a note about the UnsafeTransactionManager deprecation. It is intended to be removed, but inlining is only possible when the slices are available, and I'd rather not waste time manually doing that.

Soaryn committed:
Handles more documentation alignment and consistency.

Changes name of `onCommit` to `onRootCommit`; however, it is worth noting, we may want to consider naming this to `commitWith` and the revert be `revertTo`.

Changes `internalOpen` to just `open` in the TransactionManager as "internal" is redundant now.

Soaryn committed:
Changes `isActive` to be an instance method in TransactionContext, and renames the static method to `hasActiveTransaction` as well as adds a `lifecycle()` instanced getter method.

Renames `closeableJournals` to `journalsToCommit` to more clearly express these journals are marked as what should commit for a given depth.

Fixes some comment consistency with spaces after the `//`

Removes public access to getOpenTransaction with an arbitrary depth. Users should not be accessing other transactions that they could inadvertently commit. Instead, the only transactions they should have are the contexts or the ones they opened.

Reworks the close callbacks a little more. To make it more simplified down to just the transaction that can interact with it. This means that the only place you can interact with adding these callbacks or change what would happen during these times reside in SnapshotJournal.

Soaryn committed:
Adds basic tests for transactions. Journaling, hierarchy, and current transaction with the unsafe route.

Soaryn committed:
Changes documentation referencing SnapshotJournal `onCommit` rather than `commit`.
Removes UnsafeTransactionManager

Soaryn commented on src/main/java/net/neoforged/neoforge/transfer/transaction/TransactionContext.java:
```diff
@@ -0,0 +1,75 @@
+ * You should never need to cast this to a {@link Transaction}.
+ */
+@ApiStatus.NonExtendable
+public interface TransactionContext {
```
I wanted to quickly ask, as it has been brought up a couple times, `TransactionContext` is intended to be a limited scope of `Transaction` to prevent accidental misuse. The question: Would `TransactionView` (or similar) be possibly a clearer name for that intent? I am fine with `TransactionContext`, but it has been asked a few times as what `TransactionContext` actually is in relation to `Transaction`.

Technici4n replied:
I don't think that View is suitable as it would imply a read-only view.

Soaryn replied:
With the overall context of what this interface is, is that not what it precisely is when passing it through? A read only view of a `Transaction`. Yes, you can cast it to the full `Transaction`, but doing so would be incorrect usage wouldn't it? 🤔

Shadows-of-Fire replied:
It certainly _looks_ like a read-only view, lol

Technici4n replied:
Downcasting is not the question here. It's not read-only because you can add additional journals to it. However you are not allowed to open or close the tx.

Soaryn replied:
Well not exactly, as you only add journals to a Transaction, not the context. The context has no idea it has that sort of data.

Yes, we call `journal.updateSnapshot(transactionContext)` but that doesn't necessarily mean that the use is not read only for the vast number of consumers, we cast it to a transaction to make those addition calls. I can leave it as `context`, but I am bringing this up as I have been already asked multiple times "What is a transaction context" by people who are already familiar with what the transaction system is supposed to do.

Soaryn committed:
Removes unnecessary reverse iteration of the journals.

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/transaction/SnapshotJournalDebugInfo.java:
```diff
@@ -0,0 +1,34 @@
+// TODO: Remove after migrations have been established. This is more for info to validate assumptions on use rather than guessing and inquiring a subset.
+@Deprecated(since = ResourceHandlerDeprecationHandling.MC_1_21_6)
+@ApiStatus.Internal
+public class SnapshotJournalDebugInfo {
```
Why is this still here? I don't understand what you expect to gain from this information. Additionally, it is not thread-safe.

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/transaction/TransactionContext.java:
```diff
@@ -0,0 +1,75 @@
+     * 
+     * @throws IllegalStateException If this function is not called on the thread this transaction was opened in.
+     */
+    @ApiStatus.NonExtendable
```
Don't add `NonExtendable` here, it's already on the interface. And this whole method should be removed anyway, one should simply not get access to a non-active transaction.

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/transaction/TransactionContext.java:
```diff
@@ -0,0 +1,75 @@
+     * 
+     * @throws IllegalStateException If this function is not called on the thread this transaction was opened in.
+     */
+    Lifecycle lifecycle();
```
Lifecycle should not be an instance method, only a static method. It's fine to keep an instance method in the implementations but make it package-private.

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/transaction/TransactionManager.java:
```diff
@@ -0,0 +1,122 @@
+    private static final ThreadLocal<TransactionManager> MANAGERS = ThreadLocal.withInitial(TransactionManager::new);
+    final Thread thread = Thread.currentThread();
+    final List<Transaction> stack = new ArrayList<>();
+    final List<SnapshotJournal<?>> journalsToCommit = new ArrayList<>();
```
Put `root` somewhere in the name of the variable to make it clear that this is only the case for **root** commits and not for nested commits.

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/transaction/TransactionManager.java:
```diff
@@ -0,0 +1,122 @@
+    final List<SnapshotJournal<?>> journalsToCommit = new ArrayList<>();
+    int currentDepth = -1;
+
+    private final Int2ObjectMap<Class<?>> debugMap = new Int2ObjectOpenHashMap<>();
```
This would probably make more sense inside each transaction rather than as a map in the transaction manager.

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/transaction/snapshots/NotifyingSnapshotJournal.java:
```diff
@@ -0,0 +1,84 @@
+import net.neoforged.neoforge.transfer.transaction.TransactionContext;
+import org.jetbrains.annotations.Nullable;
+
+// TODO notes for reviewers: This is used in at least 4 different places in Neoforge implementations later, and use roughly the same layout.
```
Remove it. The other slices have not been reviewed yet so don't use them to justify this existing. If we need it later it can be readded later.

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/transaction/snapshots/StatelessSnapshotJournal.java:
```diff
@@ -0,0 +1,56 @@
+ * 
+ * @see NotifyingSnapshotJournal
+ */
+public abstract class StatelessSnapshotJournal extends SnapshotJournal<StatelessSnapshotJournal.Singleton> {
```
Remove as well.

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/transaction/SnapshotJournal.java:
```diff
@@ -0,0 +1,153 @@
+     * Return a new <b>nonnull</b> object containing the current state of this journal.
+     * <b>{@code null} may not be returned, or an exception will be thrown!</b>
+     */
+    @ApiStatus.OverrideOnly
```
`OverrideOnly` should not be used on `protected` methods, since the Java language already ensures that they are not called from the outside.

Soaryn replied:
That is not a guarantee when you want them to be subclassed. You can access protected methods without issue within subclasses, so we should make the contract to not call them with the annotation.

Technici4n replied:
You can only call them from the subclass, which is fine. `OverrideOnly` has a specific meaning (don't call this public method) and it's not the right meaning here.

Soaryn replied:
I'm not convinced I have been clear:
> `OverrideOnly` has a specific meaning (don't call this public method)

This is exactly what we **should** specify as for a user it would be essentially seen as public with how it is interacted with. Let's take a practical example:
```java
public class MyHandler<R> extends IResourceHandler<R> { // Some handler class
    SubClassedJournal journal = new SubClassedJournal(); // some journal

    public int insert(int index, R resource, int amount, TransactionContext transaction) {
        //Problem: Neither of these methods should be called by an implementer, but they do have access
        journal.createSnapshot(); // This method is present since it is abstract, which means we now have access to it here. 
        journal.onRootCommit(2); // Less likely, but still problematic. Protected in the sub class makes it reachable here due to how java works.
        
        //Correct:
        journal.updateSnapshots(transaction)
        ...
    }

    public static class SubClassedJournal extends SnapshotJournal<Integer>{

        @Override
        protected Integer createSnapshot() {
            ...
        }

        @Override
        protected void revertToSnapshot(Integer snapshot) {
            ...
        }

        @Override
        protected void onRootCommit(Integer originalState) {
            super.onRootCommit(originalState);
        }
    }
}
```

Since this is how most users are expected to make and interact with their journals, we should have something that informs them, "Hey, this is not meant to be called only overridden". Albeit the annotation does basically nothing for the IDE to highlight it seems, but the documented contract is at least present.

Saying that the annotation "should not be used on protected methods, since the Java language already ensures that they are not called from the outside" with how we are wanting a user to interact with this is fairly incorrect is it not? I'll defer to the other maintainers, but this would be an ideal scenario of the `OverrideOnly` annotation.

Technici4n replied:
We should avoid using annotations unless there is significant added value. In this case there isn't. If someone wants to call the method from their subclass it's fine. Except for the onClose method which should likely be made package-private.

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/transaction/SnapshotJournal.java:
```diff
@@ -0,0 +1,153 @@
+            Objects.requireNonNull(snapshot, "Snapshot may not be null!");
+            snapshots.set(currentDepth, snapshot);
+
+            // This is a special case where we need to cast to access the add journalToCommit method.
+            // This should not be used as a usage example for your implementations
+            // with exception to the updateSnapshots method though you should just be able to call super.updateSnapshots to avoid
+            // doing it yourself when necessary.
+            if (transaction instanceof Transaction resolvedTransaction)
+                resolvedTransaction.addClosingJournal(this);
```
Remove the useless comment, and write a cast instead of an instanceof:

```suggestion
            ((Transaction) transaction).addClosingJournal(this);
```

Soaryn replied:
We should inform someone looking at this as to WHY we do this and why they should not, as we tell them not to, then blatantly do so since it is necessary. I've at least attempted to simplify the comment.

 As for the casting, it makes no difference really, I changed it, but I am more curious as to why?

Technici4n replied:
Your updated comment is also fine. Essentially I'd rather crash if someone ever sends a custom TransactionContext implementation, rather than silently do nothing.

Soaryn replied:
We could (and likely should) seal the `TransactionContext` permitting `Transaction`

Technici4n replied:
Yeah we should.

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/transaction/SnapshotJournal.java:
```diff
@@ -0,0 +1,153 @@
+     * @throws IllegalStateException when trying to open a new transaction during this method as the current transaction is still in the process of closing.
+     */
+    @ApiStatus.OverrideOnly
+    protected void onRootCommit(T originalState) {}
```
Since this method is very often overridden, I would move it just after `releaseSnapshot` so that it comes before the huge `onClose`.

Technici4n commented on tests/src/junit/java/net/neoforged/neoforge/unittest/transfer/TransactionTests.java:
```diff
@@ -0,0 +1,93 @@
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestMethodOrder;
+
+@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
```
Why is the test order relevant here? It doesn't seem to be, so I'd remove the annotation.

Technici4n reviewed (changes requested):
OK, this is getting close to mergeable from my side. But some things still need to change.

Soaryn committed:
Moves caller source class to a field on Transaction rather than an int map on TransactionManager. We leave it as a class since we only need to resolve it as a string when it would crash in specific scenarios.

Removes annotation for test order. Originally it was there as intellij was seemingly having problems picking up it was a test, but this seems to have resolved itself at some point. (It is highly possible it was still indexing when I was first writing it)

Removes snapshot templates to look at in a later slice.

Removes transaction debug metric info due to concerns of thread safety; however, we should likely look into something at some point that we can use to debug this information quickly, to validate our expectations of how large a transaction depth can get.

Changes cast operation to a direct cast in SnapshotJournal.

Soaryn committed:
Makes `lifeCycle()` package private in Transaction as it was missed in last commit.

Moves Lifecycle enum to Transaction rather than Context since context has no methods related to its life cycle now.

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/transaction/TransactionContext.java:
```diff
@@ -0,0 +1,25 @@
+ * A subset of a {@link Transaction} that lets journals properly record, manage their state,
+ * or open inner transactions, but does not allow them to close the transaction they are passed.
+ * <p>
+ * You should never need to cast this to a {@link Transaction}.
```
Remove this comment, and instead link to SnapshotJournal. Especially to its updateSnapshots method.

Soaryn replied:
Would it be enough to use ` * @see SnapshotJournal#updateSnapshots(TransactionContext)` or did you have something else in mind?

Technici4n replied:
So I'm thinking that someone finding this class (or receiving it as argument) won't necessarily know how to use it to add transaction state. Maybe `To add state to the transaction, implement a subclass of {@link SnapshotJournal} and call its {@link SnapshotJournal#updateSnapshots} method.` or similar.

Soaryn committed:
Seals `TransactionContext` permitting only `Transaction` to prevent custom implementations from being used. These would have otherwise crashed when used with a normal flow of the transaction system.

Renames SnapshotJournal's `onClose` to `close` and makes it package private. Since this is no longer a user facing callback, it should be more aligned with being a directive.

Removes `OverrideOnly` annotations for methods in SnapshotJournal.

Adds two new tests for simultaneous transaction validations one for root, one for shared parent.

covers1624 commented on src/main/java/net/neoforged/neoforge/transfer/transaction/TransactionContext.java:
```diff
@@ -0,0 +1,28 @@
+ *
+ * @see SnapshotJournal#updateSnapshots(TransactionContext)
+ */
+@ApiStatus.NonExtendable
```
Is this annotation required now that the class is sealed?

Matyrobbrt replied:
It shouldn't be, much like final classes don't need the annotation. It's intended to be used as a warning when attempting to inherit from members that are only inheritable due to language constraints (pre-sealed or when not all possible implementations are known, etc.). But that's not the case here as there's now a compiler constraint in place.

Soaryn replied:
Woops that was supposed to be removed. My bad!

covers1624 reviewed (changes requested):
Just this, otherwise looks good.

Soaryn committed:
Removes annotation on TransactionContext that should have been removed when sealing the interface.
